rules/cache/cache.go

```diff
@@ -0,0 +1,378 @@
+	}
+
+	c.wgWorker.Add(numOngoingTasks)
+	go c.evict()
+	go c.delete()
+
+	return c
+}
+
+func (c *cache) Get(namespace string, id []byte) mrules.MatchResult {
+	c.RLock()
+	res, found := c.tryGetWithLock(namespace, id, dontSetIfNotFound)
+	c.RUnlock()
+	if found {
+		return res
+	}
+
+	c.Lock()
+	res, _ = c.tryGetWithLock(namespace, id, setIfNotFound)
+	c.Unlock()
+
+	return res
+}
+
+func (c *cache) Set(namespace string, source rules.Source) {
+	c.Lock()
+	results, exist := c.namespaces[namespace]
+	if !exist {
+		results = newResults(source)
+	} else {
+		results.source = source
+	}
+	c.namespaces[namespace] = results
+	c.Unlock()
+}
+
+func (c *cache) Delete(namespace string) {
+	c.Lock()
+	results, exists := c.namespaces[namespace]
+	if !exists {
+		c.Unlock()
+		return
+	}
+	delete(c.namespaces, namespace)
+	c.toDelete = append(c.toDelete, results.elems)
+	c.Unlock()
+
+	c.notifyDeletion()
+}
+
+func (c *cache) Close() error {
+	c.Lock()
+	if c.closed {
+		c.Unlock()
+		return errCacheClosed
+	}
+	c.closed = true
+	c.Unlock()
+
+	close(c.closedCh)
+	c.wgWorker.Wait()
+	return nil
+}
+
+// tryGetWithLock attempts to get the match result, returning true if a match
+// result is successfully determined and no further processing is required,
+// and false otherwise.
+func (c *cache) tryGetWithLock(
+	namespace string,
+	id []byte,
+	setType setType,
+) (mrules.MatchResult, bool) {
+	var res mrules.MatchResult
+	results, exists := c.namespaces[namespace]
+	if !exists {
+		return res, true
+	}
+	source := results.source
+	idHash := mid.HashFn(id)
+	elem, exists := results.elems[idHash]
+	if !exists {
+		if source == nil {
+			return res, true
+		}
+		// fall through if the result is not cached and we have a valid source.
+	} else {
+		res = elem.result
+		if source == nil || source.IsValid(res.Version()) {
+			// NB(xichen): in order to avoid the overhead acquiring an exclusive
+			// lock to perform a promotion to move the element to the front of the
+			// list, we set an expiry time for each promotion and do not perform
+			// another promotion if the previous one is still fresh. This should be
+			// good enough because if the cache is sufficiently large, the frequently
+			// accessed items should be still near the front of the list. Additionally,
+			// we can still achieve the exact LRU semantics by setting fresh duration
+			// and stutter duration to 0.
+			now := c.nowFn()
+			if elem.ShouldExpire(now) {
+				c.promote(now, elem)
+			}
+			return res, true
+		}
+		// fall through if the cached result is invalid and we have a valid source.
+	}
+	if setType == dontSetIfNotFound {
+		return res, false
+	}
+	// NB(xichen): we are guaranteed to have a valid source when we get here, and
+	// the result is either not cached, or cached but invalid, in both cases we
+	// should use the source to compute the result and set it the cache if needed.
+	return c.setWithLock(namespace, id, idHash, results, exists), true
```

Comment:
Can we simplify these logic gates?  Seems like a reasonable amount of of falling through/hard to follow paths.  Potentially some early exits might hurt with less indentation?